File: ExtractFeature/flowFeature/FlowFeature.go
package flowFeature

type FlowFeature struct {
	TCPBaseFeature
	//TCP连接基本特征（共9种）
	//Duration      float64
	//ProtocolType  string
	//Service       uint8
	//Flag          uint8
	//SrcBytes      uint
	//DstBytes      uint
	//Land          uint8
	//WrongFragment uint8
	//Urgent        uint8

	TCPContentFeature
	// TCP连接的内容特征（共13种） TCPContentFeature
	//Host             uint8
	//NumFailedLogins  uint8
	//LoggedIn         uint8
	//NumCompromised   uint
	//RootShell        uint8
	//SuAttempted      uint8
	//NumRoot          uint
	//NumFileCreations uint8
	//NumShells        uint8
	//NumAccessFiles   uint8
	//NumOutboundCmds  uint
	//IsHotLogin       uint8
	//IsGuestLogin     uint8

	TimeFlowFeature
	//基于时间的网络流量统计特征 （共9种，23～31）TimeFlowFeature
	//Count uint16
	//SrvCount uint16
	//SErrorRate float32
	//SrvSErrorRate float32
	//RErrorRate float32
	//SrvRErrorRate float32
	//SameSrvRate float32
	//DiffSrvRate float32
	//SrvDiffHostRate float32

	HostFlowFeature
	//基于主机的网络流量统计特征 （共10种，32～41）HostFlowFeature
	//DstHostCount           uint16
	//DstHostSrvCount        uint16
	//DstHostSameSrvRate     float32
	//DstHostDiffSrvRate     float32
	//DstHostSameSrcPortRate float32
	//DstHostSrvDiffHostRate float32
	//DstHostSErrorRate      float32
	//DstHostSrvSErrorRate   float32
	//DstHostRErrorRate      float32
	//DstHostSrvRErrorRate   float32
}

func (f *FlowFeature) FeatureToString() string {
	data := ""
	data += f.TCPBaseFeature.FeatureToString()
	data += f.TCPContentFeature.FeatureToString()
	data += f.TimeFlowFeature.FeatureToString()
	data += f.HostFlowFeature.FeatureToString()
	return data
}

func (f *FlowFeature) SetTCPBaseFeature(tcpBaseFeature *TCPBaseFeature) {
	f.TCPBaseFeature = *tcpBaseFeature
}

func (f *FlowFeature) SetTCPContentFeature(contentFeature *TCPContentFeature) {
	f.TCPContentFeature = *contentFeature
}

func (f *FlowFeature) SetTimeFlowFeature(timeFlowFeature *TimeFlowFeature) {
	f.TimeFlowFeature = *timeFlowFeature
}

func (f *FlowFeature) SetHostFlowFeature(hostFlowFeature *HostFlowFeature) {
	f.HostFlowFeature = *hostFlowFeature
}

func NewFlowFeature() *FlowFeature {
	return &FlowFeature{}
}

File: PredictFeature/CallPredict/flowFeature.py
# coding=utf-8
import numpy as np
from GenerateModel import type as AttackType

class FlowFeature:
    def __init__(self, request):
        self.duration = request.duration
        self.protocolType = request.protocolType
        self.service = request.service
        self.flag = request.flag
        self.srcBytes = request.srcBytes
        self.dstBytes = request.dstBytes
        self.land = request.land
        self.wrongFragment = request.wrongFragment
        self.urgent = request.urgent

        self.host = request.host
        self.numFailedLogins = request.numFailedLogins
        self.loggedIn = request.loggedIn
        self.numCompromised = request.numCompromised
        self.rootShell = request.rootShell
        self.suAttempted = request.suAttempted
        self.numRoot = request.numRoot
        self.numFileCreations = request.numFileCreations
        self.numShells = request.numShells
        self.numAccessFiles = request.numAccessFiles
        self.numOutboundCmds = request.numOutboundCmds
        self.isHotLogin = request.isHotLogin
        self.isGuestLogin = request.isGuestLogin

        self.count = request.count
        self.srvCount = request.srvCount
        self.sErrorRate = request.sErrorRate
        self.srvSErrorRate = request.srvSErrorRate
        self.rErrorRate = request.rErrorRate
        self.srvRErrorRate = request.srvRErrorRate
        self.sameSrvRate = request.sameSrvRate
        self.diffSrvRate = request.diffSrvRate
        self.srvDiffHostRate = request.srvDiffHostRate

        self.dstHostCount = request.dstHostCount
        self.dstHostSrvCount = request.dstHostSrvCount
        self.dstHostSameSrvRate = request.dstHostSameSrvRate
        self.dstHostDiffSrvRate = request.dstHostDiffSrvRate
        self.dstHostSameSrcPortRate = request.dstHostSameSrcPortRate
        self.dstHostSrvDiffHostRate = request.dstHostSrvDiffHostRate
        self.dstHostSErrorRate = request.dstHostSErrorRate
        self.dstHostSrvSErrorRate = request.dstHostSrvSErrorRate
        self.dstHostRErrorRate = request.dstHostRErrorRate
        self.dstHostSrvRErrorRate = request.dstHostSrvRErrorRate

    def toNpArray(self):
        # array = [[self.duration, self.protocolType, self.service, self.flag, self.srcBytes, self.dstBytes, self.land,
        #          self.wrongFragment, self.urgent, self.host, self.numFailedLogins, self.loggedIn, self.numCompromised,
        #          self.rootShell, self.suAttempted, self.numRoot, self.numFileCreations, self.numShells,
        #          self.numAccessFiles, self.numOutboundCmds, self.isHotLogin, self.isGuestLogin, self.count,
        #          self.srvCount, self.sErrorRate, self.srvSErrorRate, self.rErrorRate, self.srvRErrorRate,
        #          self.sameSrvRate, self.diffSrvRate, self.srvDiffHostRate, self.dstHostCount, self.dstHostSrvCount,
        #          self.dstHostSameSrvRate, self.dstHostDiffSrvRate, self.dstHostSameSrcPortRate,
        #          self.dstHostSrvDiffHostRate, self.dstHostSErrorRate, self.dstHostSrvSErrorRate, self.dstHostRErrorRate,
        #          self.dstHostSrvRErrorRate]]
        #array = [[self.protocolType, self.service, self.flag, self.srcBytes,
        #          self.sameSrvRate, self.dstHostSrvCount, self.dstHostSameSrvRate,
        #          self.dstHostDiffSrvRate, self.dstHostSrvSErrorRate]]
        array = [[self.duration, self.protocolType, self.service, AttackType.Flag.Type[self.flag], self.srcBytes,self.dstBytes,
                    self.srvSErrorRate,self.srvRErrorRate, self.sameSrvRate, self.diffSrvRate,
                    self.dstHostSameSrvRate, self.dstHostDiffSrvRate, self.dstHostSrvSErrorRate,self.dstHostSrvRErrorRate]]
        return np.array(array)
